(RK-23) Update for the use of PMT metadata library

andersonmills opened the pull request:
Before this commit, R10K used it's own metadata parsing algorithms for
modules from the Forge. After this commit, we are using shared code from
Puppet to parse the metadata files.

WARNING: This commit has two big problems that will be resolved in future PRs.
1 - Semantic should be a gem which R10K depends upon. That should be
handled soon.
2 - The shared code from puppet is currently in lib/shared/puppet.

andersonmills committed:
(RK-23) Use PMT metdata.json library

Before this commit, R10K used it's own metadata parsing algorithms for
modules from the Forge. After this commit, we are using shared code from
Puppet to parse the metadata files.

WARNING: This commit has two big problems that should be resolved in
future commits.
1 - Semantic should be a gem which R10K depends upon. That should be
handled soon.
2 - The shared code from puppet is currently in lib/shared/puppet.

adrienthebo commented on lib/r10k/module/metadata_file.rb:
```diff
@@ -0,0 +1,31 @@
+  # Attempt to read the metadata file and return the Puppet::ModuleTool::Metadata object
+  def read(metadata_file_path = @metadata_file_path)
+    if self.exist?
+      File.open(metadata_file_path) do |f|
```
This can use `@metadata_file_path.open` instead of `File.open(path)` since `@metadata_file_path` is a Pathname object.

andersonmills replied:
Happy to change this one.

adrienthebo commented on lib/r10k/module/metadata_file.rb:
```diff
@@ -0,0 +1,31 @@
+          metadata.update(JSON.load(f))
+        rescue JSON::ParserError
+          #raise ArgumentError, "Could not parse JSON #{metadata_path}", ex.backtrace
+          false
```
This seems to indicate that `#read` has a return value, if so it would be good to document that.

adrienthebo replied:
Looking more closely at the previous code indicates that this might just be a remnant of that, feel free to remove the `false`.

andersonmills replied:
Removed. forge.rb is currently the only place reading Metadata. That will be reworked in CODEMGMT-5, which is up next.

adrienthebo commented on lib/r10k/module/metadata_file.rb:
```diff
@@ -0,0 +1,31 @@
+          #returns Metadata.self
+          metadata.update(JSON.load(f))
+        rescue JSON::ParserError
+          #raise ArgumentError, "Could not parse JSON #{metadata_path}", ex.backtrace
```
This is swallowing an exception, so it might be good to log a message at debug or debug1 to indicate that the load failed.

andersonmills replied:
Happy to log. Chose debug.

adrienthebo commented on lib/semantic.rb:
```diff
@@ -0,0 +1,7 @@
+module Semantic
+  autoload :Version, 'semantic/version'
```
Since this is Semantic code this is an FYI and not a "this needs to be fixed", but Matz indicated that autoload is [deprecated](https://www.ruby-forum.com/topic/3036681) and should be avoided.

andersonmills replied:
Noted and added to personal notes about PE-3118.

adrienthebo commented on lib/r10k/module/forge.rb:
```diff
@@ -61,10 +63,9 @@ def expected_version
   # @return [R10K::SemVer] The version of the currently installed module
   def current_version
-    @metadata.read
+    #@metadata.read(@path + 'metadata.json') #Why would it not be already read?
```
https://github.com/puppetlabs/r10k/commit/a20870fdd551e7b0cfb9b3c23a0cb42393c4f201 - the metadata file would not automatically be read when instantiated, which meant that the version field would be nil until that read. Not great behavior, and I think that's an artifact of how I implemented the metadata file. As long as the metadata fields will be automatically populated when we read this, we can omit the explicit read.

andersonmills replied:
Gotcha. Certainly for the forge class here, metadata is now read on initialization.

adrienthebo commented on lib/r10k/module/forge.rb:
```diff
@@ -89,7 +90,7 @@ def status
       # The module is not installed
       return :absent
-    elsif not @metadata.exist?
+    elsif not File.exist?(@path + 'metadata.json')
```
Since `@path + 'metadata.json'` will create a Pathname object, we could do `(@path + 'metadata.json').exist?` or create the object and invoke exist on it. Stylistic change, do what you think is more readable.

andersonmills replied:
I'm still learning when Ruby-ism is correct and when it's over done, but the calling a method on the object result of an addition feels a little weird.

adrienthebo replied:
Fair point, disregard this comment.

andersonmills committed:
(RK-23) Switch off dependency merging for metadata

Before this commit, when the metadata was checked on an already
installed module, the metadata code would merge dependencies listed in
the metadata file. This commit adds a flag to not check these dependencies.
